### PR TITLE
Add Sharing panel to model detail pages

### PR DESCRIPTION
**Pull request by barrymcgee**
## Done

Add Sharing panel to model detail pages, showing who a model has already been shared with.

## QA

- Pull code
- Run `./run clean && ./run serve`
- Open http://0.0.0.0:8036/models
- Click on a model 
- Click on 'Share model' under the topology
- Verify the share model panel opens and lists the users the model has been shared with

## Details

Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1737,
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1739

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -60,6 +60,7 @@
     "classnames": "2.3.1",
     "clone-deep": "4.0.1",
     "d3": "6.7.0",
+    "date-fns": "^2.21.1",
```
Comment (hatched) [inline package.json]: can you remove the ^ plz

File: src/app/selectors.js [hunk 2]
```diff
@@ -141,7 +141,7 @@ const getDecodedMacaroons = (macaroons) => {
 const getModelUUIDByName = (name, modelData) => {
   let owner = null;
   let modelName = null;
-  if (name.includes("/")) {
+  if (name?.includes("/")) {
```
Comment (hatched) [inline src/app/selectors.js:144]: Why this was being called without a model name?

File: src/app/utils/utils.js [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import { URL } from "@canonical/jaaslib/lib/urls";
```
Comment (hatched) [inline src/app/utils/utils.js:1]: This doesn't look like it's used?

File: src/app/utils/utils.js [hunk 4]
```diff
@@ -403,3 +405,7 @@ export const generateRelationIconImage = (applicationName, modelStatusData) => {
   }
   return generateIconImg(applicationName, application.charm);
 };
+
+export const formatFriendlyDateToNow = (date) => {
+  return formatDistanceToNow(parseISO(date), { addSuffix: true });
```
Comment (hatched) [inline src/app/utils/utils.js:410]: Should we have this support a tooltip which has the actual value in it?

File: src/components/Aside/Aside.tsx [hunk 5]
```diff
@@ -23,7 +27,13 @@ export default function Aside({
         "is-pinned": pinned === true,
       })}
     >
-      {children}
+      {!loading ? (
+        children
+      ) : (
+        <div className="loading">
```
Comment (hatched) [inline src/components/Aside/Aside.tsx:33]: The loading indicator is only rendered in the content where data is missing so in the config case, only the left column gets a spinner and in this case, the same, the 'add a user' column would not get a spinner so this should be handled in the panel itself, not the aside.

File: src/panels/ShareModelPanel/ShareModel.test.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            <QueryParamProvider ReactRouterRoute={Route}>
+              <ShareModel />
+            </QueryParamProvider>
+          </TestRoute>
+        </Provider>
+      </MemoryRouter>
+    );
+    const firstUserCard = wrapper.find(".share-model__card").first();
+    const firstUserName = firstUserCard.find(".share-model__card-title strong");
+    expect(firstUserName.text()).toBe("eggman@external");
```
Comment (hatched) [inline src/panels/ShareModelPanel/ShareModel.test.tsx:30]: Are there multiple users here? Might be best to check that all are listed.

**Review by hatched — CHANGES_REQUESTED** [review]

File: src/app/selectors.js (hunk 2 above)
Comment (barrymcgee) [reply src/app/selectors.js:144]: The test was calling it without a model name before I set up the params provider.

File: src/app/utils/utils.js (hunk 3 above)
Comment (barrymcgee) [reply src/app/utils/utils.js:1]: I didn't explicity add that.. must have been VS Code - removing.

File: src/app/utils/utils.js (hunk 4 above)
Comment (barrymcgee) [reply src/app/utils/utils.js:410]: There's already an issue to do that [here](https://github.com/canonical-web-and-design/jaas-dashboard/issues/965).. that why I made this a common util.

File: src/panels/ShareModelPanel/ShareModel.test.tsx (hunk 6 above)
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.test.tsx:30]: In the data dump, there is only one.

File: src/panels/ShareModelPanel/ShareModel.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+export default function ShareModel() {
+  const { modelName } = useParams<EntityDetailsRoute>();
+
+  const modelStatusData: TSFixMe = useModelStatus() || null;
+
+  const users = modelStatusData?.info?.users;
+
+  const isOwner = (user: string) => {
+    return user === modelStatusData?.info["owner-tag"].split("-")[1];
```
[the remark this answers is not in the document]
Comment (barrymcgee) [reply src/panels/ShareModelPanel/ShareModel.tsx:21]: Good catch - refactored.

**Review by hatched — APPROVED** [review]

**Pull request merged**